File: content/app-interact/lab/support/password-cracker/python/password-cracker-multithread.py
# SPDX-License-Identifier: BSD-3-Clause

import hashlib
import itertools
from threading import Thread
from typing import List

password_hash = (
    "59a5abc2a99b95be73c31ea272ab0f2f2fe42fec30367155cb73f6f6cef1f4e6"
    "ee37f586cbd02cc738a87a5d6add3ba31dbeaf39ec77cad910837c94c65837fb"
)

password_len = 4

charset = "abcdefghijklmnopqrstuvwxyz"


def worker(first_char, result):
    for p in itertools.product(charset, repeat=password_len - 1):
        password = first_char + "".join(p)

        if hashlib.sha512(password.encode()).hexdigest() == password_hash:
            result.append(password)
            return

    result.append(None)


if __name__ == "__main__":
    threads = []
    results = []

    for i in range(len(charset)):
        res: List[str] = []
        t = Thread(target=worker, args=(charset[i], res))
        t.start()

        threads.append(t)
        results.append(res)

    for i in range(len(charset)):
        threads[i].join()

    for i in range(len(charset)):
        result = results[i][0]

        if result:
            print(f"worker {i} found {result}")

File: content/app-interact/lab/support/password-cracker/password-cracker-multithread.c
/* SPDX-License-Identifier: BSD-3-Clause */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <openssl/sha.h>
#include <pthread.h>

#include "../utils/utils.h"

#define NUM_WORKERS 26
#define PASSWORD_LEN 4
#define PASSWORD_HASH "\x59\xa5\xab\xc2\xa9\x9b\x95\xbe"		\
	"\x73\xc3\x1e\xa2\x72\xab\x0f\x2f"				\
	"\x2f\xe4\x2f\xec\x30\x36\x71\x55"				\
	"\xcb\x73\xf6\xf6\xce\xf1\xf4\xe6"				\
	"\xee\x37\xf5\x86\xcb\xd0\x2c\xc7"				\
	"\x38\xa8\x7a\x5d\x6a\xdd\x3b\xa3"				\
	"\x1d\xbe\xaf\x39\xec\x77\xca\xd9"				\
	"\x10\x83\x7c\x94\xc6\x58\x37\xfb"
#define HASH_LEN 64

int check_password(char *password, int len)
{
	unsigned char hash[HASH_LEN];

	SHA512((unsigned char *)password, len, hash);

	return memcmp(hash, PASSWORD_HASH, HASH_LEN) == 0;
}

void *worker(void *arg)
{
	char first_char;
	char password[PASSWORD_LEN+1];
	int i, k;
	int found = 0;

	first_char = (char)arg;

	password[0] = first_char;

	for (i = 1; i < PASSWORD_LEN; i++)
		password[i] = 'a'-1;

	/*
	 * Generate all possible combinations.
	 */
	k = 1;

	while (k >= 1) {
		if (password[k] < 'z') {
			password[k]++;

			if (k < PASSWORD_LEN) {
				k++;
				password[k] = 'a'-1;
			}
		}

		if (k == PASSWORD_LEN) {
			/* Check one combination. */
			if (check_password(password, PASSWORD_LEN)) {
				found = 1;
				break;
			}
		}

		if ((password[k] == 'z') || (k == PASSWORD_LEN)) {
			k--;
		}
	}

	if (found) {
		char *p;

		password[PASSWORD_LEN] = 0;
		p = strdup(password);

		return p;
	} else {
		return NULL;
	}
}

void create_workers(pthread_t *tids)
{
	int ret;
	int i;
	char char_list[] = "abcdefghijklmnopqrstuvwxyz";

	for (i = 0; i < NUM_WORKERS; i++) {
		ret = pthread_create(&tids[i], NULL, worker, (void *)char_list[i]);
		DIE(ret, "pthread_create");
	}
}

int main()
{
	pthread_t tids[NUM_WORKERS];
	int i;
	int ret;
	char *password;

	create_workers(tids);

	for (i = 0; i < NUM_WORKERS; i++) {
		ret = pthread_join(tids[i], (void **)&password);
		DIE(ret < 0, "pthread_join");

		if (password) {
			printf("worker %d found %s\n", i, password);
			free(password);
		}
	}

	return 0;
}
